File: Src/Data/Entities/PlaylistUserPermission.cs
using System;

using Microsoft.EntityFrameworkCore;

namespace MusicPlayerBackend.Data.Entities;

public enum PlaylistPermission
{
    Full = 0,
    AllowedToModifyTracks = 1,
    AllowedToView = 2
}

[PrimaryKey(nameof(PlaylistId), nameof(UserId), nameof(Permission))]
public sealed record PlaylistUserPermission : EntityBase
{
    public Guid PlaylistId { get; set; }
    public Guid UserId { get; set; }

    public PlaylistPermission Permission { get; set; }

    public Playlist Playlist { get; set; } = null!;
    public User User { get; set; } = null!;
}

File: Src/Persistence/Entities/PlaylistUserPermission.fs
namespace MusicPlayerBackend.Persistence.Entities

open System
open MusicPlayerBackend.Persistence.Entities

type PlaylistPermission =
    | Full = 0
    | ModifyTrack = 1
    | View = 2

type PlaylistUserPermission = {
    PlaylistId: PlaylistId
    UserId: UserId

    CreatedAt: DateTimeOffset
    UpdatedAt: DateTimeOffset voption
}
